### PR TITLE
Update ASR tests

### PR DESCRIPTION
**Pull request by blisc**
Changes:

- Removed the unclassified tag and tests/unclassified folder

test_unclassified_asr.py -> test_asr_gradient_step_and_eval.py

- test_jasper_training, test_quartznet_training, test_stft_conv_training now test that loss decreases after 3 steps and are marked as integration tests
- test_double_jasper_training removed as the use case is not ever encountered during use
- test_garnet_training removed as the model is now in experimental / unsupported
- test_jasper_evaluation marked as integration and skipduringci. Since Jenkins checks the ASR eval pass, not needed during CI

test_unclassified_speech_commands.py -> test_speechcommands_gradient_step_and_eval.py

- test_jasper_training renamed to test_quartznet_speech_commands_training. Now marked as integration and tests that loss decreases
- test_double_jasper_training removed
- test_stft_conv removed
- test_jasper_eval renamed to test_quartznet_speech_commands_eval and now marked as integration

test_unclassified_zeroDS_jasper.py removed

To-Dos:
- [x] Fix test_unclassified_asr.py
- [x] Fix test_unclassified_speech_commands.py
- [x] Fix test_unclassified_zeroDS_jasper.py

### CODE REVIEW

File: tests/integration/test_asr_gradient_step_and_eval.py [hunk 1]
```diff
@@ -136,82 +135,24 @@ def test_jasper_training(self):
...
-        log_probs1 = jasper_decoder1(encoder_output=encoded1)
-        log_probs2 = jasper_decoder2(encoder_output=encoded2)
-        # log_probs = mx_max1(x1=log_probs1, x2=log_probs2)
-        # encoded_len = mx_max2(x1=encoded_len1, x2=encoded_len2)
-        log_probs = log_probs1
-        encoded_len = encoded_len1
-        loss = ctc_loss(
-            log_probs=log_probs, targets=transcript, input_length=encoded_len, target_length=transcript_len,
+        self.nf.train(
+            [loss], callbacks=[callback], optimizer="sgd", optimization_params={"max_steps": 3, "lr": 0.003},
```
Comment (tkornuta-nvidia) [inline tests/integration/test_asr_gradient_step_and_eval.py:144]: 🏆 👍

File: tests/integration/test_asr_gradient_step_and_eval.py [hunk 2]
```diff
@@ -287,81 +232,26 @@ def test_stft_conv_training(self):
...
-            [loss], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
-        )
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
 
-    @pytest.mark.unclassified
+    @pytest.mark.integration
+    @pytest.mark.skipduringci
     def test_jasper_evaluation(self):
+        # Note this test still has no asserts, but rather checks that the current eval path works
```
Comment (tkornuta-nvidia) [inline tests/integration/test_asr_gradient_step_and_eval.py:251]: Ok, so when do we want this test to be executed?

During nightly build? Weekly build?

How about introducing a new marker: convergence?

I define Convergence Tests as a subclass of Acceptance tests, aiming at making sure that models that are crucial for us from business perspective (e.g. JASPER as our flagship model) or scientific perspective (e.g. in our tutorials we report scores for TRADE on MultiWOZ, so we need to make sure that it still works) converge.

@blisc What do you think about that?

**Comment by tkornuta-nvidia** [comment]
Hi, this is really great work @blisc !

Cleaner code, simpler and faster tests, nice!

**Comment by tkornuta-nvidia** [comment]
And regarding zeroDS_jasper.py: do we really need that one?

Generally I was thinking about building around zeroDS a whole "suite" of tests, just to test whether connections are working for different graphs - load a graph, configure zeroDS so it will mimic DL required for a given graph, check neural types, make forward and backward steps.

But in case of JASPER we are already covering it very thoroughly, so isn't this test redundant?

File: tests/integration/test_asr_gradient_step_and_eval.py (hunk 2 above)
Comment (blisc) [reply tests/integration/test_asr_gradient_step_and_eval.py:251]: This is not a convergence test as in weights are random. While this test creates a Jasper graph, it mainly checks the following which are not currently covered by any other pytest test:

- EvaluatorCallback
- NerualModuleFactory.eval

To me, this would be useful for users to test after installing NeMo but prior to model training to verify that their installation worked as intended. I have marked it as skipduringci as it is redundant with our Jasper Jenkins CI tests, and thus we do not need to test this as part of our testing setup.

I am unsure what the difference between an integration and an acceptance test is. To me, I am integrating the callback with our eval function to ensure that it works. Do you think this test is an acceptance test instead?

**Comment by blisc** [comment]
> And regarding zeroDS_jasper.py: do we really need that one?
> 
> Generally I was thinking about building around zeroDS a whole "suite" of tests, just to test whether connections are working for different graphs - load a graph, configure zeroDS so it will mimic DL required for a given graph, check neural types, make forward and backward steps.
> 
> But in case of JASPER we are already covering it very thoroughly, so isn't this test redundant?

I've just started looking at this. I think you're right. I should demote it to unittest and ensure that the ZerosDL output zeros. But this test seems nonsensical?

File: tests/integration/test_asr_gradient_step_and_eval.py [hunk 3]
```diff
@@ -136,82 +135,24 @@ def test_jasper_training(self):
...
-        optimizer = nemo.backends.pytorch.actions.PtActions()
-        optimizer.train(
-            [loss], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
-        )
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
 
-    @pytest.mark.unclassified
+    @pytest.mark.integration
     def test_quartznet_training(self):
```
Comment (tkornuta-nvidia) [inline tests/integration/test_asr_gradient_step_and_eval.py:152]: Hi, is this the test that samples a batch = whole set, what in turn results in lowering the loss at each step?

Can you please add a comment and describe what is happening? - this might be really useful for other people in writing their tests...

File: tests/integration/test_asr_gradient_step_and_eval.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply tests/integration/test_asr_gradient_step_and_eval.py:251]: 1) eval is tested in system tests
2) I have a "gut feeling" that during those tests we should not call evaluation and rely only on training loss - faster, the same result (we are not testing the generalization, right?)

So what you are saying clearly indicates that NeMo lacks a test suite for callbacks, maybe as part of Callbacks redesign/refactor we will take care of it, for now let's leave it the way you implemented it, those are good changes!! 👍

File: tests/integration/test_asr_gradient_step_and_eval.py (hunk 3 above)
Comment (blisc) [reply tests/integration/test_asr_gradient_step_and_eval.py:152]: Yes, all tests in that file do batch gradient descent instead of stochastic gradient descent.

I can add a comment, do you want a doc string for the tests to explain what they're testing? Or do you want me to explain how I coded my test?

Comment (tkornuta-nvidia) [reply tests/integration/test_asr_gradient_step_and_eval.py:152]: I won't stop you from adding any additional comments within the code ;) (I actually do it as I treat is a good practice), but doc strings will be sufficient

Comment (blisc) [reply tests/integration/test_asr_gradient_step_and_eval.py:164]: Added docstrings for asr and tts integration tests.
